### ISSUE TITLE
Modernise Node/SimpleNode - Request for comments

### ISSUE THREAD

**Issue by nickl-** [phax/ParserGeneratorCC#16]
Wondering if anyone else shares in my grief? I feel that if I have to so much as look at another snippet like this I might just scream.

So 1990...

```java
if ( node.jjtGetNumChildren() > 0 )
    ((SimpleNode) node.jjtGetChild(0)).evaluateNode();
```

Concerned for my own well being I decided its best I do something about it. Instead of just fixing it for my project I thought this might be something others will appreciate too. So I am starting this discussion to hear if you are interested to see this applied upstream and also to give everyone a chance to give their input towards an optimal design.

These are some possible areas to be considered, looking forward to hear what you think.

  1. Non breaking.
  2. Performance
  3. Accessibility

## 1. Non breaking.

Even though the `jjt` prefix for establishing a name space is awfully primitive it is not my desire to break backwards compatibility with this patch. I imagine it would also be preferable to retain a binary compatibility with javacc fork as well.

This dictates that the current method signatures remain unchanged and for the most part this naming convention will stay. In addition the `Node` interface contract will have to remain the same and no additional abstract methods can be specified or existing implementations will break.

If we really wanted to extend `Node` functionally it can still be accomplished with [default interface methods](https://docs.oracle.com/javase/tutorial/java/IandI/defaultmethods.html) but this will bump minimum java requirement to JDK 8 if I am not mistaken.

Another approach could be to have an interface that extends `Node` and have the generated `SimpleNode` implement the new super interface leaving all existing custom implementations unaffected. However this solution might be more trouble than its worth.

Alternatively the new functionality can be added to `SimpleNode` only and this might not be a trade off at all. Could `SimpleNode be seen as end user API and these enhancements are to improve end user accessibility. Where as the `Node` interface is more applicable to internal API and use by the Parser. If my assumptions are valid then perhaps we don't need to expose anything on `Node` at all.

## 2. Performance

Before going overboard and throwing in `Stream` or `Collection` types we might want to consider the performance impact. The use of a `Node[]` array to collect the children may have been an easy choice back in the day but even with our modern VMs arrays still drastically out performance any wrapper types. Sure streams bring with it parallelism but I don't believe we have the volume of child nodes to realise any of those benefits.

I am happy for children to remain an array as implemented.

## 3. Accessibility

Even though existing functionality may remain the same there is much that can be done to bring this out of the stone age and improve accessibility.

Personally I've been leaning towards a [`ListIterator`](https://docs.oracle.com/javase/10/docs/api/java/util/ListIterator.html) implementation which allows for forward as well as backward traversal with additional functionality to add, set or remove elements. Even just `hasNext()` on it's own will circumvent the need to compare the count of children which will already make my day.

The use of `ListIterator` only bumps minimum java requirement to JDK1.2.

For sake of completion I think it is only fair to also mention [`Iterator`](https://docs.oracle.com/javase/10/docs/api/java/util/Iterator.html) but aside from it being simpler to implement as it provides less functionality I cannot find adequate justification for choosing it over `ListIterator`.

We can either let `Node` extend this interface or implement it at `SimpleNode` or simply as an anonymous inner class making `SimpleNode` a facade delegating to the iterator. I do feel exposing the methods at class level will be more useful but we also have the option to encapsulate and provide an accessor method to the iterator.

The iterator implementation will manipulate the children array directly without the need to cast to a `Collection` or such.

Another useful accessibility enhancement would be a `getChildren` accessor for retrieving the `Node[]` array which can ensure it gets initialised if null. If for no other reason than to enable its use with an enhanced for loop instead of the geriatric alternative.

For the finishing touch I was considering adding helper functions wrapping `jjtGetChild` and `jjtGetParent` but implicitly cast to `SimpleNode`. This will go a long way to clean up redundant casts riddling the code.

Let me know if you have anything to add or disagree with and I will start rolling a patch for consideration.

**Comment by phax** [comment]
Currently on vacation. Will respond afterwards...

**Comment by nickl-** [comment]
The thumbs up is good enough for me... relax and enjoy the break. =)

**Comment by kaikalur** [comment]
You might want to check out the new code generator framework I implemented for the C# code generator. It almost completely opens up everything and you can pretty much generate any kind of code you want. I implemented the tokenizer in C# as a table driven one. That should give you maximum freedom to generate any kind of code. In stead of forking 7.0.3, you might be better off starting with that branch. That's my opinion - especially if you want something like Maven to use this stuff.

**Comment by nickl-** [comment]
@kaikalur ty for your comments

> You might want to check out the new code generator framework I implemented for the C# code generator. It almost completely opens up everything and you can pretty much generate any kind of code you want. I implemented the tokenizer in C# as a table driven one. That should give you maximum freedom to generate any kind of code. 

I am not quite following, could you point me to the implementation you are referring to perhaps that will shed more light. What it sounds like is that you are suggesting an alternative approach all together for generating code.

Perhaps I need to elaborate, I am looking at refactoring for the purposes of cleaning up the code of an existing implementation, in particular that of [beanshell/beanshell](/beanshell/beanshell) Due to the limited accessibility exposed by the existing `Node` interface the result tends to be overly verbose and repetitive. 

By my analysis with a few carefully selected method extractions extending the existing `SimpleNode` functionality it will reduce unnecessary complexity which will improve the overall legibility of the current implementation.

Since I am doing this anyway, I thought perhaps others might also find it useful so instead of just customising our `SimpleNode` class we could apply these enhancements to `SimpleNode.template` instead.

In other words, I am looking at simplifying access to the `Node` hierarchy and offer to pay the efforts forward if you think it might be of use.

> In stead of forking 7.0.3, you might be better off starting with that branch. That's my opinion - especially if you want something like Maven to use this stuff.

Not my fork nor do I want to get involved in your politics and this is certainly not the place to get into a discussion about this. The problem might already be worse than you realise. it wasn't a choice to start using this fork if anything Maven is to blame, I only noticed after the fact that we're not using javacc anymore. I could probably make an effort to revert back to 7.0.3 but it's not high on my things to do list.

You don't have to tell me that forking is a bad idea, I just spent 6 months merging 2 forks that got away from each other over the span of 5 years. Not fun!!!! It would be best for everyone if you guys would settle your differences, sooner than later and get back on one track. Rather work together on a common goal then pulling in opposite directions. Which is the spirit of open source after all, we're not only sharing the product but share in it's progression as well.

I extended this offer to you too javacc/javacc#50 with the thought that if both projects agree that this is beneficial I would only roll against javacc in hopes that it would motivate this fork to rebase from upstream again. In so doing my little part to help keep the trees closer together. =) 

Have you taken the time to talk about merging your efforts at all?

**Comment by kaikalur** [comment]
Oh, may be I need to give some context here. First of all - there are no politics :) here. I just think that a lot of fixes to the codegenerators are adhoc and generally involve messing with a bunch of files including some in the core part of the product. That plus I need to generate C# in my most recent job at MSFT so I started factoring out the code generator *completely* from the main core of javacc in such a way that the  core takes the grammar file and produces a data structure. Then I also supplied a RI for a driver that can use that code and driver to generate a table-driven token manager. Parser is a bit more involved. So for now I just factored it out of the core, but not table driven yet. You can see that code in the following branch:

https://github.com/javacc/javacc/tree/csharp_codegen

And if you guys adopt to java, that will be awesome because that we will also get a modern parser for Java - may be even use java8 or java9! That's my vested interest. And if someone can do C++ that will be even more awesome.

The advantage of this approach you actually absolutely need no knowledge of javacc/jjtree/parser etc. You can just blindly translate the C# code generator into Java using proper java idioms etc. And if someone has really specific kind of code to be generated (say some places don't allow you to use exceptions), you should be able to to it by just writing the code generator (or taking an existing one and hacking it).

And IMO all the old effforts render obsolete with this new thing as it's a completely new way of doing things. I'm hoping I will find some time to merge my stuff and actually release JavaCC 8.0. But for that we need both Java and C++ code generators using the new framework. I'm too busy with other stuff so I'm not able to get it it. I would happy to help if someone wants to take it on.

Hope that helps.

Sreeni.

**Comment by kaikalur** [comment]
Sorry I meant to say - look at the codegenerate/csharp folder for a codegenerator that does not need to touch the javacc code.

**Comment by nickl-** [comment]
Ok I see so we have refactored the code generation which makes it easier to implement for different languages or special requirements. 

But from what I can see it is still only going to generate SimpleNode children.  Even though the templates for Node and SimpleNode are now combined in a  [single template](https://github.com/javacc/javacc/blob/csharp_codegen/codegenerator/csharp/src/resources/templates/csharp/Node.template) it still only exposes the ever so basic `jjtGetNumChildren` and `jjtGetChild`.

Which does not help me at all. 

To get me vested at the very least I require a boolean check for the availability of children and child traversal which does not require me keeping a counter. There also seems to be something missing from `Node`, I am just not sure if there is a generic solution to solve the problem. Basically, if I have to constantly cast to `SimpleNode` then why am I working with `Node` types to begin with? My third deal breaker is never needing to cast to `SimpleNode`.

If you are open to improving accessibility of these base types then I will write the java refactor for you.

**Comment by kaikalur** [comment]
Like I said, there is no "it" anymore as far as code generation is considered. It's now a user option - CODE_GENERAOTR so you bring your own code generator! And there is no API per se from JavaCC itself. It's all done by the code generator where *every thing* - class/field/interface/function etc. is upto it. So I'm also thinking of moving all the weirdo codegen options like LEGACY_EXCEPTION_HANDLING etc. out of the core completely. Even things like NAMESPACE etc. don't belong in javacc options. They should all be codegen options.

So in summary, go ahead and do the right thing(TM) for modern versions of java.

**Comment by nickl-** [comment]
Okay now your previous statement "the old effforts render obsolete" make much more sense. The compiler compiler does not dictate what gets generated, if you want to generate a hierarchy of `Fruit` instead of a `Node` hierarchy then enjoy.

I don't think this will have much effect to beanshell/beanshell we are very much married to the current AST and unlikely to ever drastically change the current building blocks.

Based on this assessment I agree, what I proposed is more suited as a private mod to our custom `SimpleNode`. 

@kaikalur you may ping me if you need a hand with a clear requirement and defined deliverable, will gladly help. I can also be tasked with C++, C# or any other lingo you desire but at this very moment I have zero uptake requirements for java which may way in on my decision to defer.  For now I nub this distraction in the butt and return to the grind at hand, the modern version of java scripting and interpretation =)
